lib/spack/spack/solver/splicing.py

```diff
@@ -0,0 +1,91 @@
+# Copyright Spack Project Developers. See COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+from functools import cmp_to_key
+from typing import Dict, List
+
+import spack.deptypes as dt
+from spack.spec import Spec
+from spack.traverse import by_dag_hash, traverse_nodes
+
+
+class Splice:
+    """A class representing a single splice. These will always be indexed in a
+    dict by the parent spec that they are being spliced into. Note: this should
+    be a dataclass, but Python 3.6 dose not support dataclasses.
+
+    splice_spec (Spec): the spec being spliced into a parent
+    child_name (str): the name of the child that `splice_spec` is replacing
+    child_hash (str): the hash of the child that `splice_spec` is replacing
+    """
+
+    def __init__(self, splice_spec: Spec, child_name: str, child_hash: str):
+        self.splice_spec = splice_spec
+        self.child_name = child_name
+        self.child_hash = child_hash
+
+    def __repr__(self) -> str:
+        output = (
+            f"Splice spec {self.splice_spec}, "
+            f"replacing child {self.child_name} "
+            f"with hash {self.child_hash}"
+        )
+        return output
```

Comment:
Here we can use a `NamedTuple` to avoid encoding a `__repr__`:
```python
>>> from typing import NamedTuple
>>> class Foo(NamedTuple):
...     foo: int
...     bas: str
...     
>>> f = Foo(foo=1, bas="abcd")
>>> repr(f)
"Foo(foo=1, bas='abcd')"
```
```suggestion
class Splice(NamedTuple):
    #: The spec being spliced into a parent
    splice_spec: Spec
    #: The name of the child that splice spec is replacing
    child_name: str
    #: The hash of the child that `splice_spec` is replacing
    child_hash: str
```